[Feature] Implement passive anticheat system

ariel- opened the pull request:
**Changes proposed**: Original patch taken from here: (created by @maanuel)
https://community.trinitycore.org/topic/10-patch-telenpc2-pvpranks-passive-anticheat-for-trinitycore-2016-02-24-4e4b2b9-35a-only/
- This adds the core capabilities to passively detect player hacking/cheating using 3rd party tools, it just scans for some flags/speed mismatch and emits a warning to every GM (or rather any account with rbac::RBAC_PERM_RECEIVE_GLOBAL_GM_TEXTMESSAGE)
- The original version was massively updated, it was optimized and updated to reflect core changes along the way.

**Target branch(es)**: 335/6x

**Issues addressed**: None, feature request.

**Tests performed**: (Does it build, tested in-game, etc) It builds and correctly detects many kinds of 3rd party tools, issuing a warning.

**Known issues and TODO list**:
- [x] Move anticheat.sql to correct folder and update base sql
- [x] Move singleton out of header
- [ ] Re evaluate logic of existing checks (specially non-efficient ones)

myran2 commented on sql/updates/world/2016_mm_dd_nn_world.sql:
```diff
@@ -0,0 +1,28 @@
+(@ENTRY+10, '============================='),
+(@ENTRY+11, 'Players with the lowest averages:'),
+(@ENTRY+12, 'Player: %s Average: %f Total Reports: %u'),
+(@ENTRY+13, 'Players with the more reports:'),
```
Players with the most reports?

ariel- replied:
Heh, that's what you get for brainless copy pasting hardcoded texts xD

Nelidon commented:
You should rebase your branch and squash commits.

ariel- commented:
When it's deemed for merge into the core, that is.

Until then I prefer to leave the commit history intact so changes may be seen step by step (also different authors on commits)

Shauren commented on sql/base/characters_database.sql:
```diff
@@ -1,8 +1,8 @@
--- MySQL dump 10.13  Distrib 5.6.26, for Win64 (x86_64)
```
I see many unrelated changes in the full sql dump - charsets/unsigned on some fields

Rochet2 commented on src/server/game/Anticheat/AnticheatMgr.cpp:
```diff
@@ -0,0 +1,482 @@
+        return;
+
+    ObjectGuid::LowType lowGuid = player->GetGUID().GetCounter();
+    AnticheatData& plData = m_playerMap[lowGuid];
```
I dont think this is safe. `StartHackDetection` can be called from multiple threads at the same time and access to `m_playerMap` is not guarded by anything from multiple accesses and modifications at the same time. There may be other thread safety related issues also.

Shauren replied:
Looks like it is safe to me entry in m_playerMap is created on login (threadsafe context) and destroyed on logout (threadsafe context)
Modifying a single map value is not unsafe for m_playerMap itself but can be unsafe for the value - however this is not a problem either because only one thread can process movement packets for one player.

Rochet2 replied:
It would be better probably to change the `[]` operator to `.at`. This way we can detect in case unintended behavior occurs. And it looks safe as you say, however the code also uses `m_playerMap.clear();` in one of the commands, which in my mind makes it unsafe since all data is deleted even though ppl are online and get detected.

Shauren replied:
Oh, the existence of command changes this completely - indeed it becomes unsafe

Shauren commented on src/server/game/Anticheat/AnticheatMgr.h:
```diff
@@ -0,0 +1,153 @@
+};
+
+// GUIDLow is the key.
+typedef std::unordered_map<ObjectGuid::LowType, AnticheatData> AnticheatPlayersDataMap;
```
Please do not use lowguid as key, it should only be used in database (I know there are MANY uses for it still in the code but don't add new ones)

Keader commented:
The Hitchhiker's Hack  is detected?

Rochet2 commented on sql/updates/characters/2016_03_13_00_characters.sql:
```diff
@@ -0,0 +1,29 @@
+  `jump_reports` bigint(20) unsigned NOT NULL DEFAULT '0',
+  `waterwalk_reports` bigint(20) unsigned NOT NULL DEFAULT '0',
+  `teleportplane_reports` bigint(20) unsigned NOT NULL DEFAULT '0',
+  `climb_reports` bigint(20) unsigned NOT NULL DEFAULT '0',
```
Seems like the bigint is unnecessary as data is written in uint32, which is unsigned int. And data is never loaded, or I cant see that.

ariel- replied:
Data is queried when using .anticheat global command, however I'll have to rethink if this is necessary at all.

jackpoz commented:
I would like this patch to be tested with valgrind, helgrind and a static analyzer of your choice

Rochet2 commented on src/server/scripts/Commands/cs_anticheat.cpp:
```diff
@@ -0,0 +1,264 @@
+        if (!*args)
+            return false;
+
+        std::string strCommand = const_cast<char*>(args);
```
doesnt look like a necessary cast. The value should be copied by the string constructor.

Rochet2 commented on src/server/game/Handlers/MovementHandler.cpp:
```diff
@@ -345,6 +346,9 @@ void WorldSession::HandleMovementOpcodes(WorldPacket& recvData)
     }
 
+    if (plrMover)
+        sAnticheatMgr->StartHackDetection(plrMover, movementInfo, Opcodes(opcode));
```
The Opcodes cast looks unnecessary as opcode is already `uint16` and `StartHackDetection` takes an `uint16`

joschiwald commented on src/server/game/Anticheat/AnticheatData.h:
```diff
@@ -0,0 +1,89 @@
+
+private:
+    uint16 lastOpcode;
+    MovementInfo lastMovementInfo;
```
this is the same like `player->GetMovementInfo()`

ariel- commented:
Well, I'm open to suggestions, because I don't know about multi threaded programming. To solve the unsafe issue, then I don't have to clear the map when the command is issued right? Can it be fixed by 0'ing out the associated data instead of clearing the whole map?

And, lastMovementInfo is saved to compare with new movement, see https://github.com/TrinityCore/TrinityCore/pull/16774/files#diff-cd67866d024cb7e5288fca94a3cd3358R172

joschiwald commented:
`player->GetMovementInfo()` contains the same data as lastMovementInfo, so you don't need to save it twice

Shauren commented:
Instead of trying to make access to m_playerMap from multiple threads safe (and slow!) you should just drop the delete command entirely

Shauren commented on src/server/game/Spells/SpellEffects.cpp:
```diff
@@ -16,6 +16,7 @@
  * with this program. If not, see <http://www.gnu.org/licenses/>.
  */
 
+#include "AnticheatMgr.h"
```
there is no change in this file - include not needed

Ascathor commented on src/server/scripts/Commands/cs_anticheat.cpp:
```diff
@@ -0,0 +1,210 @@
+        static WorldLocation const jailPos(1, 16226.5f, 16403.6f, -64.5f, 3.2f);
+
+        Player* pTarget = nullptr;
+        char* command = strtok(const_cast<char*>(args), " ");
```
const_cast is overkill here, simply replace it with a strtok(*args, " ");
That's how commands were handled until now.

Rochet2 replied:
Not quite sure what you mean. Const cast isn't an overhead (probably optimized away as it is only typesafety). Also before it was still casted with a C style cast: `strtok((char*)args, " ");`
https://github.com/TrinityCore/TrinityCore/blob/478cc756eb05591f36857be3ffd1895684529710/src/server/scripts/Commands/cs_character.cpp#L956

Ascathor replied:
Ah, yeah. I forgot the actual casting that still needs to happen. But that's the case, we do not need an actual const_cast type of convertion, a simple c-style cast is how we handled it until now. It solves the issues.

ariel- replied:
How it was in the past != how it should be done.

I need to remove const-ness because strtok requires a char*.

Ascathor replied:
Are those commands different to the way other commands are handled ?

ariel- replied:
Not really, as Rochet2 pointed out in that link, const-ness is casted away as it is here (except the cast there is not explicit, rather a C-style cast, this is why I prefer const_cast, because it's explicit of my intention)

Ascathor commented on src/server/game/Anticheat/AnticheatMgr.cpp:
```diff
@@ -0,0 +1,447 @@
+
+bool FlyHackCheck::OnCheck(Player* player, MovementInfo const& /* movementInfo */, uint16 /*opcode = 0*/) const
+{
+    if (AnticheatData* playerData = GetPlayerData(player->GetGUID()))
```
Travis complains here. Unused variable

jackpoz commented on src/server/game/Anticheat/AnticheatMgr.h:
```diff
@@ -0,0 +1,151 @@
+    public:
+        static AnticheatMgr* instance()
+        {
+           static AnticheatMgr instance;
```
please keep an eye on TC development and apply the latest standard, for example https://github.com/TrinityCore/TrinityCore/commit/b7f936891aeb86eeb4eb67caa78e672858aa1e20

jackpoz commented on src/server/game/Anticheat/AnticheatMgr.h:
```diff
@@ -0,0 +1,151 @@
+        void LoadSettings();
+
+    private:
+        AnticheatData* GetPlayerData(ObjectGuid const& guid)
```
is it safe to return a pointer to a inner container element ?

Shauren replied:
Sure it is - ObjectMgr does that all the time and here its a private method so it shouldn't be a problem (unless its used in an evil way and exposed to code outside of this class)

jackpoz commented on src/server/game/Anticheat/AnticheatMgr.h:
```diff
@@ -0,0 +1,151 @@
+        virtual void HackReport(Player* player) const = 0;
+
+    protected:
+        static AnticheatData* GetPlayerData(ObjectGuid const& guid) { return sAnticheatMgr->GetPlayerData(guid); }
```
@Shauren sAnticheatMgr->GetPlayerData(guid) exposed and used in places like this one ?

ariel- replied:
It's only an accessor, used in checks context, checks are only called on movementInfo handlers, which in turn only get called when player is logged in, data is already existing at this point.

Rochet2 replied:
Assumptions. Assumptions lead to crashes. (if not now, then later)
Im sure jackpoz would feel a lot safer if it returned a copy or the pointer was wrapped in `std::unique_ptr`, right?

ariel- replied:
Wait, what assumption? Return value is always checked against nullptr, I see no reason this could crash.

jackpoz replied:
> Return value is always checked against nullptr, I see no reason this could crash.

null dereference is not really the only C++ kind of crash, especially when returning a pointer to a inner container element that can be deleted at anytime simply by calling container.clear() .

@Rochet2: I don't like any kind of std::cheat_ptr , it's still worth to note in PR review which are the points of attention, also to evaluate the quality and safety of code by reading the author replies.

Shauren replied:
How about you remove this method entirely and just pass AnticheatData* as parameter to OnCheck?

ariel- replied:
clear() was removed entirely, and passing the pointer as a parameter sounds good.

Nawuko commented on src/server/game/Anticheat/AnticheatMgr.cpp:
```diff
@@ -0,0 +1,453 @@
+    // We also must check the map because the movementFlag can be modified by the client.
+    // If we just check the flag, they could always add that flag and always skip the speed hacking detection.
+    // 369 == DEEPRUN TRAM
+    if (player->m_movementInfo.HasMovementFlag(MOVEMENTFLAG_ONTRANSPORT) && player->GetMapId() == 369)
```
Why exclude the deeprun tram here? If the gameobjects (tram) are correctly marked as Transports the flag will be applied automaticly. I don't like the idea of hardcoding exceptions like these.

Shauren replied:
The problem with this is that only deeprun tram is NOT a Transport* serverside, it is the same gameobject type as elevators what means it does not maintain a passenger list

Nawuko replied:
@Shauren It's marked as GAMEOBJECT_TYPE_TRANSPORT in the database so the client automaticly sends MOVEMENTFLAG_ONTRANSPORT. Atleast on 3.3.5. I just checked the sended movement package from the client on the latest rev.

Shauren replied:
Yes but only MO_TRANSPORT are Transport* - this kind of check/exclusion is useless if you cant verify if the player really is on the tram or not (the cheater would just need to send  ONTRANSPORT flag with any random transport gameobject guid present on map and he would be able to speedhack all he wants)

Nawuko replied:
Yes you are totally right. I tested 5 different cheats and one actually does exactly this.

Nawuko commented on src/server/game/Anticheat/AnticheatMgr.cpp:
```diff
@@ -0,0 +1,453 @@
+    //    return;
+    float x, y, z;
+    player->GetPosition(x, y, z);
+    float ground_Z = player->GetMap()->GetHeight(x, y, z);
```
This will be triggered on every movement package (even if the player only changes facing) which can be many many GetHeight queries if the player for example only spins arround. Maybe some kind of throttling is a good idea. Something like this is quite expensive if someone has 2k+ Players and every movement package is checked.

ariel- replied:
Well, the actual "hack checking" code is the only thing remnant from the original; any suggestions here, besides throttle? Maybe caching the value elsewhere (is it good idea?)

maanuel replied:
Hi everyone.

It has been long time since I wrote the original AC code. I can't really recall the idea behind that check.

I think that I could be trying not to divide by zero (in a trigonometric function?) in the GetHeight function, I don't know if that could really happen.
Another reason could be that I detected something in the  WoWEmuHacker's (for 3.3.5) behaviour.

Independently of the reasons above explained, I think that the check is not correct. GetPositionZ is a float number, so it must be checked against an epsilon.

About the performance concerns, at that moment I did not understand the TC threading model. Now I understand that anticheat checks are made in the same thread that is processing packets, so it is not a nice idea to perform heavy computations. 

I think that the options available to mitigate this drawback, if the anticheat checks are made in the same thread, are the followings:

Randomize this check, I mean it could happen with certain probability. If you are detected positive, your probability increases otherwise decreases.
Inspecting the GetHeight() function I noticed that there are two metrics used: maps data and vmaps data. We could consider just using maps data, in case it were faster than vmap computations (I don't really know about the peformance costs of each).

Finally, we could consider doing all this checks in a separated thread, in case that map or vmap height were thread safe. @Shauren Could you enlighten us about this? 

If I am not wrong vmap height is thread-safe unless you set isHeightCalcEnabled as false, but it is just set at startup.

jackpoz replied:
Checking only half of height data (only maps and not vmaps) makes no sense since you have partial informations.
Since this is just an optional feature, you could explain in the configs about the performance drawbacks that these check will do and then it's up to the server owner to choose if it's worth to enable them or not.

Nawuko commented on src/server/game/Anticheat/AnticheatMgr.cpp:
```diff
@@ -0,0 +1,453 @@
+
+bool SpeedHackCheck::OnCheck(Player* player, MovementInfo const& movementInfo, uint16 /*opcode = 0*/) const
+{
+    AnticheatData* playerData = GetPlayerData(player->GetGUID());
```
I dont think this is needed anymore (was used for the last saved movement info which is now directly used from the Player class)

Nawuko commented:
I just tested some random cheats for 3.3.5

**TeleportPlaneCheck** does in 4/5 cases not work. (Player has falling flag).

**SpeedhackCheck** dosent work in 3/5 cases (Player has OnTransportFlag or the cheat skips/obscure packages, send many many small packages or delays packages).

**WalkOnWaterCheck** fails in 5/5 cases (all cheats dosent send the waterwalk movementflag but are still able to walk on water)

jackpoz commented on src/server/game/Anticheat/AnticheatMgr.cpp:
```diff
@@ -0,0 +1,453 @@
+    SQLTransaction trans = CharacterDatabase.BeginTransaction();
+    // Mutex protected part
+    {
+        boost::shared_lock<boost::shared_mutex> lock(*HashMapHolder<Player>::GetLock());
```
@Shauren : is the lock needed here ?

Shauren replied:
If running from World::Update thread? No.

jackpoz replied:
ping on the comment above about the lock not being needed (if running from World::Update thread: is it ? )

jackpoz commented on src/server/game/Anticheat/AnticheatMgr.cpp:
```diff
@@ -0,0 +1,453 @@
+    }
+    CharacterDatabase.CommitTransaction(trans);
+
+    QueryResult resultDB = CharacterDatabase.Query("SELECT guid,average,total_reports FROM players_reports_status WHERE total_reports > 0 ORDER BY average ASC LIMIT 5");
```
are the CommitTransaction() and Query() both sync ? otherwise you have a database race condition

Shauren replied:
CommitTransaction is async, Query is sync

ariel- replied:
Yes, however it'll lead to not-updated-enough data, maybe split this command in two: anticheat save and anticheat global?

jackpoz replied:
All these commands are evil enough to allow sync DirectCommitTransaction(...) which will wait for the transaction to be over before executing the Query(...)

jackpoz commented on src/server/game/Anticheat/AnticheatMgr.h:
```diff
@@ -0,0 +1,151 @@
+            return nullptr;
+        }
+
+        AnticheatPlayersDataMap m_playerMap;
```
I always wonder if we should keep maps using Player guid as key or just store the data in Player class itself

jackpoz commented on src/server/game/Anticheat/AnticheatMgr.cpp:
```diff
@@ -0,0 +1,453 @@
+void AnticheatMgr::HandlePlayerLogout(ObjectGuid const& guid)
+{
+    // We must also delete it at logout to prevent have data of offline players in the db when we query the database (IE: The GM Command)
+    PreparedStatement* stmt = CharacterDatabase.GetPreparedStatement(CHAR_DEL_AC_PLAYER_REPORTS);
```
is there a way to avoid executing an additional query on logout for every player ?

ariel- replied:
I can leave the data in DB, and then do an async delete on login, anyways all data is wiped on server startup, so it can work.

jackpoz replied:
is this query executed even if the whole AntiCheat stuff is disabled ?

jackpoz replied:
ping on my questions above

jackpoz commented on src/server/worldserver/worldserver.conf.dist:
```diff
@@ -1321,6 +1322,55 @@ Warden.BanDuration = 86400
+#        Default:     1 - (Enabled)
+#                     0 - (Disabled)
+
+Anticheat.Enable = 1
```
Warden is disabled by default so it makes sense to keep this disabled by default too.

jackpoz commented on src/server/scripts/Commands/cs_anticheat.cpp:
```diff
@@ -0,0 +1,212 @@
+        return true;
+    }
+
+    static bool HandleAnticheatJailCommand(ChatHandler* handler, char const* args)
```
this command doesn't check any AntiCheat related data, it's just a generic .jail command, yet it requires anticheat to be enabled. The "!sWorld->getBoolConfig(CONFIG_ANTICHEAT_ENABLE)" check should be removed and the command should be moved to a more generic commands group, maybe the same of the freeze command (if it exists)

jackpoz commented on src/server/game/Anticheat/AnticheatMgr.cpp:
```diff
@@ -0,0 +1,453 @@
+
+void AnticheatMgr::HandlePlayerLogin(Player* player, PreparedQueryResult result)
+{
+    AnticheatData& plData = m_playerMap[player->GetGUID()];
```
the whole idea of this line in case result is null is to create the AnticheatData element in the container ?

ariel- replied:
Yes, indeed.

Runico commented on src/server/game/Anticheat/AnticheatMgr.cpp:
```diff
@@ -0,0 +1,453 @@
+    if (player->m_movementInfo.HasMovementFlag(MOVEMENTFLAG_ONTRANSPORT) && player->GetMapId() == 369)
+        return false;
+
+    uint32 distance2D = (uint32)movementInfo.pos.GetExactDist2d(&player->m_movementInfo.pos);
```
while player have flyable aura ( like fly mounts ) player could easily bypass this, you should check 3D distance while player have ability to fly

How to reproduce
1. mount up ( fly )
2. change your speed 20 ( by cheats tools )
3. fly as 3D ( easiest way move by mouse or press both w and q )

ryoting commented:
this can check player`s flyhack?
ago,this anticheat check gm`s fly only,can`t check player`s fly

Demonid commented:
@ryoting This is true, at least the last time i checked AC2, it only checks Flyhack for accouts with GM Level > 0

Dont know if this is the case for this PR, but if someone could give it a try :+1:

Aokromes commented:
This is not based on ac2.

Demonid commented:
@Aokromes Wouldn't hurt to check just in case.

Then again, what about adding some kind of movement restriction, kinda like this Anticheat does:

https://bitbucket.org/lordpsyan/lordpsyan-patches/src/bd5c2b08de445bafbb4cc70cba56bc391e5f5848/Individual/Anti-Cheats/2016_02_29-Non-Passive-AC2.patch?at=master&fileviewer=file-view-default

Just a suggestion, nothing more.

mthsena commented:
> @ryoting This is true, at least the last time i checked AC2, it only detects Flyhack for accounts with GM Level > 0
> 
> Dont know if this is the case for this PR, but if someone could give it a try :+1:

@Demonid, @ryoting  Related to WorldSession.cpp. On my version, I removed these lines and it worked fine for accounts 0.

```
REMOVE_VIOLATING_FLAGS(mi->HasMovementFlag(MOVEMENTFLAG_FLYING | MOVEMENTFLAG_CAN_FLY) && GetSecurity() == SEC_PLAYER &&
!GetPlayer()->m_mover->HasAuraType(SPELL_AURA_FLY) &&
!GetPlayer()->m_mover-HasAuraType(SPELL_AURA_MOD_INCREASE_MOUNTED_FLIGHT_SPEED),
MOVEMENTFLAG_FLYING | MOVEMENTFLAG_CAN_FLY);
```

Aokromes commented:
@Demonid this is intended to be PASSIVE anticheat.

Nawuko commented:
I dont want to disrespect your work. But maybe we should implement something like this: https://github.com/Nawuko/TrinityCore/commit/88d662f7b867f804629cbe1dbfb8fd2988de81d0
And not an old, only partly working code. Just an thought. Like i said i tested this anti cheat with the most popular 5 cheats out there and the result is not very great.

Aokromes commented:
@nawuko that code is under consideration.

tripleslash commented:
@Nawuko This is not supposed to be merged whatsoever. It is my personal testing playground. I plan to rewrite the entire movement system from scratch in a blizzlike way during the next semester holidays.

ryoting commented:
any news of this?

dandi2001 commented:
Is this heading anywhere? D:

Palabola commented:
Instead of try to create the perfect movement calculation on server side and looking after mismatch, my advise focus to catch Hack Engines poor designed packets and logics.

We need calculation only for speed and Z movement.

**Example:**
Some Hack Engine only spam 2 types of Movement packet(Fall & Stop Fall) with always the same flags.

tripleslash commented:
Then I'll go ahead and implement one that circumvents your suggestion. For an open source emulator a proper anticheat is necessary and not something like this.

Ghost-R commented:
@tripleslash any news ?

Aokromes commented:
@ariel- you think this can be merged? it can be enhanced  on the future.

jackpoz commented on src/server/game/Anticheat/AnticheatMgr.cpp:
```diff
@@ -0,0 +1,443 @@
+
+    Position playerPos = player->GetPosition();
+
+    float deltaZ = fabsf(playerPos.GetPositionZ() - movementInfo.pos.GetPositionZ());
```
please avoid using C functions and use the C++ std::* counterpart instead

jackpoz commented on src/server/game/World/World.cpp:
```diff
@@ -1289,6 +1290,13 @@ void World::LoadConfigSettings(bool reload)
     m_bool_configs[CONFIG_UI_QUESTLEVELS_IN_DIALOGS] = sConfigMgr->GetBoolDefault("UI.ShowQuestLevelsInDialogs", false);
 
+    // Passive anticheat
+    m_bool_configs[CONFIG_ANTICHEAT_ENABLE] = sConfigMgr->GetBoolDefault("Anticheat.Enable", true);
```
why is it enabled by default ?

Aokromes committed:
Passive anticheat optimization and code cleanup:
- Moved scripts to Custom ScriptLoader
- Migrated commands to rbac system, anticheat jail command will also apply freeze aura to target
- Oneline methods moved to header and made inline for maximum performance
- Player data is saved as unordered_map now to save on cpu cycles
- Use of "ObjectGuid::LowType" in order to keep up with core naming rules
- Added a configurable delay between reports of same Player
- Moved reports to chat window, this will also honor rbac for Receive GM messages, meaning it can be configured for gm's players accounts.

Aokromes committed:
Core/Anticheat: c fabsf to c++ std::fabs

Aokromes committed:
Merge branch '3.3.5' of https://github.com/Aokromes/TrinityCore into passive

Demonid commented:
@Aokromes  Needs update after https://github.com/TrinityCore/TrinityCore/commit/7026c7ed7a39ae258eb0ba01f6946c8bc1594ce0

Aokromes committed:
Merge branch '3.3.5' into passive

tripleslash commented:
I'd be sad to see this thing merged into the official TC master...
I'd suggest abandoning this branch and starting to fix the fundamental issues (like movement ACK handling) first. As soon as that's done one can implement the missing handlers for the movement packets and stop handling every different opcode in one handler function and start doing proper timestamp synchronization.

Nawuko commented:
@tripleslash :+1: 

Also TrinityCore gets more and more cluttered (AHBot, Custom Duel stuff, Custom PvP Token stuff) now a half-assed AntiCheat that accomplish nothing in the most cases.

Aokromes commented:
@Nawuko trinitycore was made with the mind of add custom features since the start.

jackpoz commented:
> cluttered

?

Demonid commented:
Could anyone check if this PR is still working?

I tried after the last change but it doesn't detect anything, maybe the detection system broke after all the code changes?

Thanks!

Keader commented:
i cant see problems with custom stuffs. Using "disabled" by default, why not be present?

Demonid commented:
Could anyone check if this PR is still working?

I tried testing it after the last commit 2fb0418 but it doesn't detect anything, maybe the detection system broke after all the code changes?

Thanks!

Mrgrasser commented:
Break passive ? not work now :(

Mrgrasser commented:
@Demonid  this AntiCheat have some problem for back jump , Improved Charge warrior , spell Engineering , Some time speed rouge , Aspect of the Pack hunter , Blazing Speed mage 
If the configuration is correct is with no trouble AntiCheat.

> Then again, what about adding some kind of movement restriction, kinda like this Anticheat does:
>
> https://bitbucket.org/lordpsyan/lordpsyan-patches/src/bd5c2b08de445bafbb4cc70cba56bc391e5f5848/Individual/Anti-Cheats/2016_02_29-Non-Passive-AC2.patch?at=master&fileviewer=file-view-default
>
>Just a suggestion, nothing more.

Aokromes committed:
Merge branch '3.3.5' of https://github.com/Aokromes/TrinityCore into passive

# Conflicts:
#	sql/base/auth_database.sql
#	sql/base/characters_database.sql

Aokromes committed:
delete incorrect file

chaodhib commented:
I don't think this should get merged neither to be honest. I think we should keep the good stuff (which is the anti cheat logic, basically every OnCheck method in AnticheatMgr.cpp) and throw the rest out. 

The anti cheat logic will already use additional CPU resources. I don't think adding a system that tracks cheating players and does database queries on top of that is a good idea. Also, keeping small the anti movement cheat code will make it easier to maintain and improve. Imo, we should just add a hook in the MovementHandler: It will call a method like `bool verifyClientPosition()` that takes as input the latest verified state of the player and the latest received. If it returns false, the handler should ignore the last packet, update the client to a valid position and potentially do one (or more) of the following options (configurable by config file): log the event, kick the player and or ban the player. This change won't take much more than 300 LOC.

I think the focus should be put on the validating incoming packets (example: if the unit is already falling, ignore any new MSG_MOVE_JUMP, etc...), not all the extra stuff around it.

Aokromes committed:
Merge branch '3.3.5' into passive

Aokromes committed:
Compile fix after changes

Aokromes committed:
typo fix
